lib/engine/game/g_18_esp/game.rb

```diff
@@ -860,6 +860,7 @@ def must_buy_train?(entity)
             end
           end
           trains = trains.dup.reject { |t| t.track_type == :broad } if north_corp?(entity) && !entity.interchange?
+          trains = trains.dup.reject { |t| t.track_type == :all } if combined_train_blocked?(entity)
```

Comment:
I'm not keen on all these calls to `trains.dup`. Could the entire method be rewritten to something like this?
```
        def must_buy_train?(entity)
          return false if depot.depot_trains.empty?
          
          entity.trains.none? do |train|
            next false if extra_train?(train)
            
            case train.track_type
            when :narrow
              north_corp?(entity) || (entity.interchange? && entity.type != :minor)
            when :broad
              !north_corp?(entity) || entity.interchange?
            when :all
              !combined_train_blocked?(entity)
          end
        end
```
I've probably got the logic wrong in there, but I think it would be much clearer if it looked something like this.